File: swift/HyperBabelDemo/Screens/VideoCallScreen.swift
/*
 * Video call screen — joins the active call session for a room and surfaces
 * a status line. Rendering remote video tracks would require UIKit interop
 * (UIViewRepresentable + the SDK's video canvas type); this skeleton focuses
 * on the API + token + join lifecycle that customers most often need to copy.
 */
import SwiftUI
#if canImport(AgoraRtcKit)
import AgoraRtcKit
#endif

struct VideoCallScreen: View {
    let roomId: String
    let onHangup: () -> Void

    @EnvironmentObject var session: Session
    @State private var status: String = "Connecting…"

    var body: some View {
        VStack(alignment: .leading, spacing: 16) {
            Text("Video call").font(.title2).bold()
            Text(status).foregroundStyle(.secondary)
            Spacer()
            Button(role: .destructive, action: { hangup() }) {
                Text("Hang up").frame(maxWidth: .infinity)
            }.buttonStyle(.borderedProminent)
        }
        .padding(20)
        .navigationBarBackButtonHidden(true)
        .task { await join() }
    }

    private func join() async {
        do {
            let active = try await UnitedChatService.getActiveVideoCall(roomId: roomId)
            // Named `callSession` so it does not shadow the app's `session`
            // (the EnvironmentObject that holds the signed-in user id).
            guard let callSession = active.session, let channelName = callSession.channelName else {
                status = "No active session for this room."
                return
            }
            guard let sessionId = callSession.sessionId else {
                status = "Session has no id — cannot request a token."
                return
            }
            let uid = callSession.uid ?? Int.random(in: 1...1_000_000)
            // Publisher tokens are session-scoped: the server checks that this
            // user is a participant and signs the token with the session's
            // channel + uid, which the response echoes back.
            let token = try await RtmService.rtcToken(
                RtcTokenRequest(channelName: channelName, uid: uid, role: "publisher",
                                sessionId: sessionId, externalUserId: session.userId)
            )
            HyperBabelVideo.shared.initialize(appId: token.appId, delegate: nil)
            try await HyperBabelVideo.shared.joinWithToken(
                channelName: token.channelName, rtcToken: token.rtcToken,
                uid: token.uid, role: "publisher"
            )
            status = "Joined \(token.channelName)"
        } catch {
            status = "Failed to start call: \((error as? ApiError)?.errorDescription ?? error.localizedDescription)"
        }
    }

    private func hangup() {
        HyperBabelVideo.shared.leaveCall()
        Task {
            _ = try? await UnitedChatService.leaveVideoCall(roomId: roomId, userId: session.userId)
            await MainActor.run { onHangup() }
        }
        HyperBabelVideo.shared.release()
    }
}

File: kotlin/app/src/main/kotlin/com/hyperbabel/demo/ui/screens/VideoCallScreen.kt
/*
 * Video call screen — joins the active call session for a room and surfaces
 * a status line. Rendering remote video tracks would require a SurfaceView
 * factory; this skeleton focuses on the API + token + join lifecycle that
 * customers most often need to copy.
 */
package com.hyperbabel.demo.ui.screens

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Button
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.dp
import com.hyperbabel.demo.api.ApiClient
import com.hyperbabel.demo.data.Session
import com.hyperbabel.demo.video.HyperBabelVideo
import io.agora.rtc2.IRtcEngineEventHandler as VideoEventHandler
import kotlin.random.Random
import kotlinx.coroutines.launch
import androidx.compose.runtime.rememberCoroutineScope

@Composable
fun VideoCallScreen(roomId: String, onHangup: () -> Unit) {
    val ctx = LocalContext.current
    val scope = rememberCoroutineScope()
    val video = remember { HyperBabelVideo(ctx.applicationContext) }
    var status by remember { mutableStateOf("Connecting…") }

    LaunchedEffect(roomId) {
        try {
            val active = ApiClient.unitedChat.getActiveVideoCall(roomId).session
            if (active?.channelName == null) {
                status = "No active session for this room."
                return@LaunchedEffect
            }
            // Issue a token first so we have the appId before initialising.
            val uid = active.uid ?: Random.nextInt(1, 1_000_000)
            val sessionId = active.sessionId
                ?: return@LaunchedEffect run { status = "Session has no id — cannot request a token." }
            val tok = ApiClient.rtm.rtcToken(
                com.hyperbabel.demo.data.RtcTokenRequest(
                    channelName = active.channelName,
                    uid = uid,
                    role = "publisher",
                    // Publisher tokens are session-scoped: the server checks that
                    // this user is a participant and signs the token with the
                    // session's channel + uid, which the response echoes back.
                    sessionId = sessionId,
                    externalUserId = Session.userId,
                ),
            )
            video.init(tok.appId, object : VideoEventHandler() {
                override fun onJoinChannelSuccess(channel: String?, uid: Int, elapsed: Int) {
                    status = "Joined $channel"
                }
                override fun onUserJoined(uid: Int, elapsed: Int) {
                    // Keep the publishing resolution in step with the call size —
                    // HyperBabel meters by the total resolution each participant
                    // receives (video/VideoQuality.kt).
                    video.trackRemoteJoined(uid)
                    status = "Peer joined: $uid"
                }
                override fun onUserOffline(uid: Int, reason: Int) {
                    video.trackRemoteLeft(uid)
                    status = "Peer left ($reason)"
                    onHangup()
                }
            })
            video.engineOrNull()?.joinChannel(tok.rtcToken, tok.channelName, null, tok.uid)
        } catch (t: Throwable) {
            status = "Failed to start call: ${t.message}"
        }
    }

    DisposableEffect(Unit) {
        onDispose {
            try { video.leaveCall() } catch (_: Throwable) {}
            try {
                scope.launch {
                    try {
                        ApiClient.unitedChat.leaveVideoCall(
                            roomId,
                            mapOf("user_id" to Session.userId),
                        )
                    } catch (_: Throwable) {}
                }
            } catch (_: Throwable) {}
            video.release()
        }
    }

    Column(
        modifier = Modifier.fillMaxSize().padding(16.dp),
        verticalArrangement = Arrangement.spacedBy(12.dp),
    ) {
        Text("Video call", style = MaterialTheme.typography.headlineSmall)
        Text(status, color = MaterialTheme.colorScheme.onSurfaceVariant)
        Button(onClick = onHangup) { Text("Hang up") }
    }
}
